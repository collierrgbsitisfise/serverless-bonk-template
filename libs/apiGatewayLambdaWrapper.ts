import { APIGatewayEvent, Context, Callback, APIGatewayProxyResult } from 'aws-lambda';

type Headers = { [key: string]: string };

const defaultHeaders = {
  'Content-Type': 'application/json',
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Credentials': true,
};

const onSuccesHandler = (
  data: any,
  statusCode: number,
  headers?: Headers,
): APIGatewayProxyResult | PromiseLike<APIGatewayProxyResult> => ({
  statusCode,
  headers: {
    ...defaultHeaders,
    ...headers,
  },
  body: JSON.stringify(data),
});

const onErrorHandler = async (error: Error): Promise<APIGatewayProxyResult> => {
  return {
    statusCode: 500,
    headers: defaultHeaders,
    body: JSON.stringify(error),
  };
};

export const apiGatewayLambdaWrapper = (
  lambda: (
    event: APIGatewayEvent,
    context?: Context,
    callback?: Callback,
  ) => [any, number, Headers] | Promise<[any, number, Headers]>,
  onSucces: (
    value: any,
    statusCode: number,
    headers?: Headers,
  ) => APIGatewayProxyResult | PromiseLike<APIGatewayProxyResult> = onSuccesHandler,
  onError: (error: Error) => Promise<APIGatewayProxyResult> = onErrorHandler,
) => {
  return function wrapp(event: APIGatewayEvent, context: Context, callback: Callback): Promise<APIGatewayProxyResult> {
    return Promise.resolve()
      .then(() => lambda(event, context, callback))
      .then(([res, code, headers]: [any, number, Headers]) => onSucces(res, code, headers))
      .catch(onError);
  };
};
