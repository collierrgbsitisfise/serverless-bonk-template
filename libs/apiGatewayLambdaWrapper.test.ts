import { APIGatewayEvent, Context, Callback } from 'aws-lambda';

import { apiGatewayLambdaWrapper } from './apiGatewayLambdaWrapper';

describe('apiGatewayLambdaWrapper', () => {
  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
  });

  it('should call "onSuccesHandler"', async () => {
    const fakeEvent = {} as APIGatewayEvent;
    const fakeContext = {} as Context;
    const fakeCallBack = jest.fn() as Callback;

    const lambdaResponse = [{}, 200, {}];
    const successHandler = jest.fn();
    const errorHandler = jest.fn();
    const lambda = jest.fn().mockReturnValueOnce(lambdaResponse);

    const executable = apiGatewayLambdaWrapper(lambda, successHandler, errorHandler);
    expect.assertions(4);
    const result = await executable(fakeEvent, fakeContext, fakeCallBack);

    expect(result).toMatchSnapshot();
    expect(lambda).toBeCalledWith(fakeEvent, fakeContext, fakeCallBack);
    expect(successHandler).toBeCalledWith(...lambdaResponse);
    expect(errorHandler).not.toBeCalled();
  });

  it('should call "onErrorHandlerr"', async () => {
    const fakeEvent = {} as APIGatewayEvent;
    const fakeContext = {} as Context;
    const fakeCallBack = jest.fn() as Callback;
    const fakeError = 'error message';

    const successHandler = jest.fn();
    const errorHandler = jest.fn();
    const lambda = jest.fn().mockRejectedValueOnce('error message');

    const executable = apiGatewayLambdaWrapper(lambda, successHandler, errorHandler);
    expect.assertions(4);
    const result = await executable(fakeEvent, fakeContext, fakeCallBack);

    expect(result).toMatchSnapshot();
    expect(lambda).toBeCalledWith(fakeEvent, fakeContext, fakeCallBack);
    expect(successHandler).not.toBeCalled();
    expect(errorHandler).toBeCalledWith(fakeError);
  });
});
