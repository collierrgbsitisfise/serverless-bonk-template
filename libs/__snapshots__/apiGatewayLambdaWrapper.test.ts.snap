// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiGatewayLambdaWrapper should call "onErrorHandlerr" 1`] = `undefined`;

exports[`apiGatewayLambdaWrapper should call "onSuccesHandler" 1`] = `undefined`;
